src/sknnr/transformers/_ccora.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+from scipy.stats import f
+
+
+def ftest_cor(p, q, N, cor):
+    """ "Direct port of yaImpute ftest_cor in yai"""
```

Comment:
I agree with everything you've noted here, including the steps to be compliant.  If we did want to avoid having to use the GPL v3 license, I'm not totally clear how we would re-implement everything from scratch.  There probably aren't too many ways to do the same thing (at least not well) and one might still use the original work as inspiration.

I don't know how much you've been using (or trust!) ChatGPT, but I found the following Q & A pretty interesting:

> Q. when you use functions from another python package in your code, do you have to abide by their licensing restrictions?  For example, if you use a library that is licensed as GPL v3, can your project be BSD v3?
> 
> A. When you use functions or libraries from another Python package in your code, the licensing restrictions of the library you are using generally apply to your project as well. This means that if you use a library licensed under the GPL v3, your project needs to comply with the terms of the GPL v3 license.
> 
> The GPL is a copyleft license that requires any derivative works or projects that use GPL-licensed code to also be licensed under the GPL or a compatible license. This "viral" nature of the GPL ensures that the freedoms and obligations of the GPL license are preserved in derivative works.
> 
> If you want to license your project under a different license, such as the BSD v3 license, and you are using a GPL-licensed library, there could be a conflict. The GPL v3 is generally not compatible with the BSD v3 license because it imposes more stringent requirements on derivative works.
> 
> To address this, you have a few options:
> 
> 1. Use a Compatible Library: Look for alternative libraries that are licensed under a more permissive license, such as the BSD v3 license. This way, you can avoid the licensing conflict altogether.
> 
> 2. Dual Licensing: Contact the library's author or copyright holder and inquire about the possibility of obtaining a separate license for your use case. They may grant you a special license or provide an alternative license option that is compatible with your desired licensing terms.
> 
> 3. Separate Execution: If your project can be structured in a way that the GPL-licensed library is a separate component and is not directly linked to or integrated into your code, you may be able to avoid the requirement of applying the GPL to your entire project. In this case, the GPL-licensed code would be considered a separate program, and the licensing obligations would apply only when distributing or modifying the GPL-licensed component.
> 
> It's important to consult the specific terms and conditions of the library's license to understand your obligations. Licensing issues can be complex, so seeking legal advice or discussing with the library's maintainers or a licensing expert can provide further clarity in navigating these situations.
> 
So, you've mentioned the possibility of 1 which makes sense, although I believe the world of kNN software is pretty narrow!  I could probably get Andy or Nick (`yaImpute` developers) to allow us to have a dual license (2), but they are probably bound by the ANN library and couldn't release it in that way.  I think 3 is a pretty interesting approach, although much of what we're doing is more or less a port of `yaImpute`, so I don't know if that would be worth the effort.

Sorry, just more questions than answers!

Reply:
> I don't know how much you've been using (or trust!) ChatGPT

I was definitely consulting with Google Bard trying to figure out how to comply with these licenses!

>  I could probably get Andy or Nick (yaImpute developers) to allow us to have a dual license (2), but they are probably bound by the ANN library and couldn't release it in that way.

Another interesting facet here is that the [ANN library](http://www.cs.umd.edu/~mount/ANN/) is distributed under [GNU **Lesser** Public License v2](http://www.cs.umd.edu/~mount/ANN/Files/1.1.2/License_1.1.2.txt), which I believe allows you to distribute either under LGPL or a more restrictive GPL. So maybe we could get a dual license to avoid GPL v3, but would still be bound by LGPL, which probably isn't worth the effort...

Reply:
I'm going to email Nick Crookston and Andy Finley on this to see what they might recommend.  My impression is that the GPL license on yaImpute came after they were actively involved.  Whether that was because the new maintainer sought to make it GPL v3 or if they were trying to be compliant, I'm not sure.  

That's pretty interesting about ANN.  I think we both concur that we'd like to be a bit less restrictive than GPL v3.  Any word I get back from Nick or Andy, I'll post in #29.